src/contexts/players-context.tsx

```diff
@@ -43,42 +46,134 @@ export interface PlayerType {
+      if (
+        patch[key] &&
+        typeof patch[key] === "object" &&
+        !Array.isArray(patch[key]) &&
+        Object.keys(patch[key]).every((input: any) => {
+          if (typeof input === "number") {
+            return Number.isInteger(input);
+          } else if (typeof input === "string") {
+            const num = Number(input);
+            return Number.isInteger(num) && input.trim() === num.toString();
+          }
+          return false;
+        })
+      ) {
+        for (const arrIndex in patch[key]) {
+          newPatch[key][arrIndex] = normalizePatch(
+            patch[key][arrIndex],
+            target[key][arrIndex],
+            true,
+          );
+        }
+      } else {
+        // If the patch is a non-object, replace the target array with the patch.
+        newPatch[key] = patch[key];
+      }
+    } else {
+      // Recursively normalize nested objects.
+      newPatch[key] = normalizePatch(patch[key], target[key], inArray);
+    }
+  }
+
+  // If we are in an array, ensure that missing fields in the patch are filled from the target.
+  if (inArray) {
+    Object.keys(target).forEach((field) => {
+      if (!(field in newPatch)) {
+        newPatch[field] = target[field];
+      }
+    });
+  }
+
+  return newPatch;
 }
 
-export function mergeDeep(target: any, ...sources: any[]) {
+/**
+ * Recursively merges properties from source objects into a target object, creating a new object.
+ * This function does not mutate the original target but returns a new object.
+ * It only updates references within the new object when there are actual changes to content or children,
+ * regardless of the depth of those changes. Arrays are copied rather than merged, and nested objects
+ * are recursively populated. This function can handle an arbitrary number of source objects.
+ * @param target The initial object to merge properties into.
+ * @param sources One or more objects from which properties will be sourced.
+ * @returns The target object merged with properties from all source objects.
+ */
+export function mergeDeep(target: any, ...sources: any[]): any {
+  const isObject = (item: any) => item && typeof item === "object";
+
   if (!sources.length) return target;
   const source = sources.shift();
+  const newTarget = Array.isArray(target) ? [...target] : { ...target };
 
   if (isObject(target) && isObject(source)) {
     for (const key in source) {
       if (isObject(source[key])) {
-        if (!target[key]) Object.assign(target, { [key]: {} });
-        mergeDeep(target[key], source[key]);
+        if (!target[key]) {
+          newTarget[key] = Array.isArray(source[key]) ? [] : {};
+        }
+        newTarget[key] = mergeDeep(newTarget[key], source[key]);
       } else {
-        Object.assign(target, { [key]: source[key] });
+        newTarget[key] = source[key];
       }
     }
   }
-
-  return mergeDeep(target, ...sources);
+  return mergeDeep(newTarget, ...sources);
```

Comment:
Don't use ambiguous variable names, harder to keep track of when going back to it.

Reply:
+1